cmd/tether/attach.go

```diff
@@ -0,0 +1,369 @@
+var (
+	Signals = map[ssh.Signal]int{
+		ssh.SIGABRT: 6,
+		ssh.SIGALRM: 14,
+		ssh.SIGFPE:  8,
+		ssh.SIGHUP:  1,
+		ssh.SIGILL:  4,
+		ssh.SIGINT:  2,
+		ssh.SIGKILL: 9,
+		ssh.SIGPIPE: 13,
+		ssh.SIGQUIT: 3,
+		ssh.SIGSEGV: 11,
+		ssh.SIGTERM: 15,
+		ssh.SIGUSR1: 10,
+		ssh.SIGUSR2: 12,
+	}
+)
+
+// PtyRequestMsg the RFC4254 struct
+type ptyRequestMsg struct {
+	Term     string
+	Columns  uint32
+	Rows     uint32
+	Width    uint32
+	Height   uint32
+	Modelist string
+}
+
+// WindowChangeMsg the RFC4254 struct
+type WindowChangeMsg struct {
+	Columns  uint32
+	Rows     uint32
+	WidthPx  uint32
+	HeightPx uint32
+}
+
+type signalMsg struct {
+	Signal string
+}
+
+var server attachServer
+
+type attachServer interface {
+	start() error
+	stop()
+}
+
+// conn is held directly as it's how we stop the attach server
+type attachServerSSH struct {
+	conn   *net.Conn
+	config *ssh.ServerConfig
+
+	enabled bool
+}
+
+// start is not thread safe with stop
+func (t *attachServerSSH) start() error {
+	if t == nil {
+		return errors.New("attach server is not configured")
+	}
+
+	if t.enabled {
+		return nil
+	}
+
+	// don't assume that the key hasn't changed
+	pkey, err := ssh.ParsePrivateKey(config.Key)
+	if err != nil {
+		detail := fmt.Sprintf("failed to load key for attach: %s", err)
+		log.Error(detail)
+		return errors.New(detail)
+	}
+
+	// An SSH server is represented by a ServerConfig, which holds
+	// certificate details and handles authentication of ServerConns.
+	// TODO: update this with generated credentials for the appliance
+	t.config = &ssh.ServerConfig{
+		PublicKeyCallback: func(c ssh.ConnMetadata, key ssh.PublicKey) (*ssh.Permissions, error) {
+			if c.User() == "daemon" {
+				return &ssh.Permissions{}, nil
+			}
+			return nil, fmt.Errorf("expected daemon user")
+		},
+		PasswordCallback: func(c ssh.ConnMetadata, pass []byte) (*ssh.Permissions, error) {
+			if c.User() == "daemon" {
+				return &ssh.Permissions{}, nil
+			}
+			return nil, fmt.Errorf("expected daemon user")
+		},
+		NoClientAuth: true,
+	}
+	t.config.AddHostKey(pkey)
+
+	t.enabled = true
+	go t.run()
+
+	return nil
+}
+
+// stop is not thread safe with start
+func (t *attachServerSSH) stop() {
+	if t == nil || !t.enabled {
+		return
+	}
+
+	t.enabled = false
+	if t.conn != nil {
+		(*t.conn).Close()
+		t.conn = nil
+	}
+}
+
+// run should not be called directly, but via start
+// run will establish an ssh server listening on the backchannel
+func (t *attachServerSSH) run() error {
+	var sConn *ssh.ServerConn
+	var chans <-chan ssh.NewChannel
+	var reqs <-chan *ssh.Request
+	var err error
+
+	// keep waiting for the connection to establish
+	for t.enabled && sConn == nil {
+		// wait for backchannel to establish
+		conn, err := utils.backchannel(context.Background())
+		t.conn = &conn
+
+		// create the SSH server
+		sConn, chans, reqs, err = ssh.NewServerConn(*t.conn, t.config)
+		if err != nil {
+			detail := fmt.Sprintf("failed to establish ssh handshake: %s", err)
```

Comment:
Should the conn be closed in this case or does the framework handle that for us here?